README.md
toonapi
=======

Node.js package for Eneco/Quby Toon API (unofficial)

[![npm](https://img.shields.io/npm/v/toonapi.svg?maxAge=300)](https://github.com/fvdm/nodejs-toonapi/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-toonapi.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-toonapi)
[![Dependency Status](https://gemnasium.com/badges/github.com/fvdm/nodejs-toonapi.svg)](https://gemnasium.com/github.com/fvdm/nodejs-toonapi#runtime-dependencies)
[![Coverage Status](https://coveralls.io/repos/github/fvdm/nodejs-toonapi/badge.svg?branch=master)](https://coveralls.io/github/fvdm/nodejs-toonapi?branch=master)

* [Node.js](https://nodejs.org)
* [Package on npm](https://www.npmjs.com/package/toonapi)
* [Toon API](https://www.toonapi.com)
* [API documentation](https://www.toonapi.com/documentation)


Beta notice
-----------

This package is not ready for production while
it is in the v0.x.x version range.
Once it's released as v1.0.0 or higher it is stable for production uses.


Installation
------------

`npm i toonapi --save`


License
-------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


Author
------

[Franklin van de Meent](https://frankl.in)


CHANGELOG.md
#### 0.3.1 (2016-7-12)

##### Bug Fixes

* **errors:** Catch all three kinds of xml errors ([6ce10d71](https://github.com/fvdm/nodejs-toonapi/commit/6ce10d71e1f7e285ddc3325c1c3e65de68002d02))

### 0.3.0 (2016-7-12)

##### Documentation Changes

* **readme:** Added link to npm package ([053cf96d](https://github.com/fvdm/nodejs-toonapi/commit/053cf96d1d3d195df768fe11c90afb4e7324d37a))

##### New Features

* **request:** Get accessToken when needed ([670d7e64](https://github.com/fvdm/nodejs-toonapi/commit/670d7e647cab709a249f1cc11c611a9d1436b85c))

##### Bug Fixes

* **errors:** Renamed message var ([86e4c7e1](https://github.com/fvdm/nodejs-toonapi/commit/86e4c7e16292d990ca7181ee2ca2c55a778ebb60))
* **main:** Catch all xml API errors ([050ce45b](https://github.com/fvdm/nodejs-toonapi/commit/050ce45b90d1a863040a7204f4ecb398a1bebdd9))
* **response:**
  * Default empty body to obj.success true ([3f9dbc63](https://github.com/fvdm/nodejs-toonapi/commit/3f9dbc630abce6a725087993a32762cfcf377ef2))
  * Default empty body to empty object ([71aad0fc](https://github.com/fvdm/nodejs-toonapi/commit/71aad0fcb0101bfc1f2d998ec3b99fffbbbf33df))

##### Tests

* **main:**
  * Check data.success value ([abe01f92](https://github.com/fvdm/nodejs-toonapi/commit/abe01f926d957ea1175f914d6adf3851b1889b41))
  * Added method devices.update ([ac13328a](https://github.com/fvdm/nodejs-toonapi/commit/ac13328a90e1e7b4050991f3f1e52ecec4b15fb5))
  * Run oauth.getTokenFromPassword after other methods ([09496cab](https://github.com/fvdm/nodejs-toonapi/commit/09496cabe9d6ac6ea6d43372cdfcd47df7fdbad5))
  * Fixed missing cache object ([80b2deb6](https://github.com/fvdm/nodejs-toonapi/commit/80b2deb6b010040f9f0f1da1d56358690b056e46))
  * Fixed minor typos ([52781c48](https://github.com/fvdm/nodejs-toonapi/commit/52781c482e78d4674553865f3b8131521ed764da))
  * Added method devices.device.get test ([4a5a5f07](https://github.com/fvdm/nodejs-toonapi/commit/4a5a5f07e5895327498701deb9373fd28c1131bb))
  * Check device uuid, cache list ([68ed9610](https://github.com/fvdm/nodejs-toonapi/commit/68ed96101c01024e8a9b968281d4c0dfd4c3470b))
  * Fixed wrong method name in last commit ([16e2d9b6](https://github.com/fvdm/nodejs-toonapi/commit/16e2d9b66d8efc9e3883b82229ee882ba4ec7189))
  * Added method devices.list test ([6e77a6ca](https://github.com/fvdm/nodejs-toonapi/commit/6e77a6ca3aae0a8da292ba4aacc2926e0ef54877))

### 0.2.0 (2016-7-9)

##### Chores

* **cleanup:** Removed trailing spaces ([2fc11a97](https://github.com/fvdm/nodejs-toonapi/commit/2fc11a973cfa7d90efccfd2d3538f82da454acc5))
* **methods:** Renamed method function names ([54062313](https://github.com/fvdm/nodejs-toonapi/commit/54062313abc412a3810db64922f22571dcedf17d))

##### Documentation Changes

* **readme:**
  * Added beta notice ([ffa2dc27](https://github.com/fvdm/nodejs-toonapi/commit/ffa2dc27d964dd30f08c4965b46be66209fd8bd3))
  * Added basic readme ([f2486de4](https://github.com/fvdm/nodejs-toonapi/commit/f2486de4702b2cfc470e5913363305153a6b88d5))

##### New Features

* **errors:** Added error endpoint missing ([d19659a4](https://github.com/fvdm/nodejs-toonapi/commit/d19659a4ebf5e215de2961129f41745dd671c4df))
* **oauth:** Added method getTokenFromPassword ([5252d4ba](https://github.com/fvdm/nodejs-toonapi/commit/5252d4ba993625f34d03e5eaef7be1b808eb5d6f))
* **setup:** Added username and password settings ([d1d2e921](https://github.com/fvdm/nodejs-toonapi/commit/d1d2e9219ebbe66d24142c0e24ed014d9764fb3c))

##### Bug Fixes

* **pushEvent:** Fixed bad request path ([93e8b81c](https://github.com/fvdm/nodejs-toonapi/commit/93e8b81cb1152bf9e6b3c8368c50a0b3606b2160))
* **devices:** Fixed bad request path ([56c9d986](https://github.com/fvdm/nodejs-toonapi/commit/56c9d986b6300924d129d95cf8ccb24c52ddbf84))
* **consumption:** Fixed bad request path ([1411fca1](https://github.com/fvdm/nodejs-toonapi/commit/1411fca1a09e6b8b6ac93a572affe8ca5af87952))
* **temperature:** Fixed request paths ([cb07a91b](https://github.com/fvdm/nodejs-toonapi/commit/cb07a91b76369bebfc7f34b92bddb6880eff2a86))
* **errors:** Fixed missing body for invalid response ([3de07430](https://github.com/fvdm/nodejs-toonapi/commit/3de07430f789275fc9b14e36fc0acf96737161a1))
* **display:** Fixed status API path ([1bc3088c](https://github.com/fvdm/nodejs-toonapi/commit/1bc3088c62204b3221581501e7f15486ca83c97e))
* **oauth:** Disable Authorization header ([35ff3a9b](https://github.com/fvdm/nodejs-toonapi/commit/35ff3a9b77d2fdcbdd9857ae6fb5d468c1118ab5))
* **response:**
  * Added /revoke handling ([a8d9fe9f](https://github.com/fvdm/nodejs-toonapi/commit/a8d9fe9fd5e775df01a50b83a517eb8edc78349e))
  * Parse XML error from API ([046f931b](https://github.com/fvdm/nodejs-toonapi/commit/046f931b05709e36aa36317269cdbde8a8b5e0c3))
* **setup:** Fixed fatal error when object is missing ([3b4930d6](https://github.com/fvdm/nodejs-toonapi/commit/3b4930d6a323196474cc5f8366770c3521aa1c21))
* **methods:** Fixed agreements API path ([6e57de0d](https://github.com/fvdm/nodejs-toonapi/commit/6e57de0d5a5cfe55ff9788ee9fe372cd3dd3a3bb))
* **reponse:** Consistent return ([a03c3ead](https://github.com/fvdm/nodejs-toonapi/commit/a03c3ead066e2b8ac0b38c5c126338b21f2841f3))

##### Refactors

* **main:**
  * Moved methods to app object ([9ffdfa6a](https://github.com/fvdm/nodejs-toonapi/commit/9ffdfa6a7b5b523033addae5ccfa18ae3a17cb31))
  * Removed unavailable agreements.update method ([fdb595ba](https://github.com/fvdm/nodejs-toonapi/commit/fdb595ba1f8c4ed97e4ba9cd1da04957ef47c381))
  * Cleaner object props assignment ([152dd2d7](https://github.com/fvdm/nodejs-toonapi/commit/152dd2d728c78cb5f98c0c53eb361cbafa15099e))
* **temperature:** Removed method temperature.update ([e8d9b02d](https://github.com/fvdm/nodejs-toonapi/commit/e8d9b02d04884f6a5ef7ae08d2fe407e3b11cab6))
* **errors:**
  * Removed xml <ams:fault> parsing ([e59c14b5](https://github.com/fvdm/nodejs-toonapi/commit/e59c14b580601fd294c5482ef99d4fe5b845a914))

test.js
var dotest = require ('dotest');
var cache = {};

var config = {
  accessToken: process.env.ACCESSTOKEN,
  clientId: process.env.CLIENT_ID,
  clientSecret: process.env.CLIENT_SECRET,
  password: process.env.CLIENT_PASSWORD,
  redirectUri: process.env.REDIRECT_URI,
  refreshToken: process.env.REFRESHTOKEN,
  timeout: process.env.TIMEOUT,
  username: process.env.CLIENT_USERNAME
};

var app = require ('./');
var toonapi = app (config);


dotest.add ('Module interface', function (test) {
  var agreements = toonapi && toonapi.agreements;

  var consumption = toonapi && toonapi.consumption;
  var electricity = consumption && consumption.electricity;
  var districtheat = consumption && consumption.districtheat;
  var gas = consumption && consumption.gas;

  var devices = toonapi && toonapi.devices;
  var device = devices && devices.device;

  var display = toonapi && toonapi.display;
  var oauth = toonapi && toonapi.oauth;
  var pushEvent = toonapi && toonapi.pushEvent;

  var temperature = toonapi && toonapi.temperature;
  var states = temperature && temperature.states;

  test ()
    .isFunction ('fail', 'exports', app)
    .isObject ('fail', 'interface', toonapi)

    .isObject ('fail', '.agreements', agreements)
    .isFunction ('fail', '.agreements.list', agreements && agreements.list)

    .isObject ('fail', '.consumption', consumption)
    .isObject ('fail', '.consumption.electricity', electricity)
    .isFunction ('fail', '.consumption.electricity.data', electricity && electricity.data)
    .isFunction ('fail', '.consumption.electricity.flows', electricity && electricity.flows)
    .isObject ('fail', '.consumption.districtheat', districtheat)
    .isFunction ('fail', '.consumption.districtheat.data', districtheat && districtheat.data)
    .isObject ('fail', '.consumption.gas', gas)
    .isFunction ('fail', '.consumption.gas.data', gas && gas.data)
    .isFunction ('fail', '.consumption.gas.flows', gas && gas.flows)

    .isObject ('fail', '.devices', devices)
    .isObject ('fail', '.devices.device', device)
    .isFunction ('fail', '.devices.device.get', device && device.get)
    .isFunction ('fail', '.devices.device.update', device && device.update)
    .isFunction ('fail', '.devices.device.data', device && device.data)
    .isFunction ('fail', '.devices.device.flows', device && device.flows)
    .isFunction ('fail', '.devices.list', devices && devices.list)
    .isFunction ('fail', '.devices.update', devices && devices.update)

    .isObject ('fail', '.display', display)
    .isFunction ('fail', '.display.status', display && display.status)

    .isObject ('fail', '.oauth', oauth)
    .isFunction ('fail', '.oauth.getTokenFromPassword', oauth && oauth.getTokenFromPassword)
    .isFunction ('fail', '.oauth.refreshToken', oauth && oauth.refreshToken)
    .isFunction ('fail', '.oauth.revokeToken', oauth && oauth.revokeToken)

    .isObject ('fail', '.pushEvent', pushEvent)
    .isFunction ('fail', '.pushEvent.subscribe', pushEvent && pushEvent.subscribe)
    .isFunction ('fail', '.pushEvent.unsubscribe', pushEvent && pushEvent.unsubscribe)

    .isObject ('fail', '.temperature', temperature)
    .isObject ('fail', '.temperature.states', states)
    .isFunction ('fail', '.temperature.states.list', states && states.list)
    .isFunction ('fail', '.temperature.states.update', states && states.update)
    .done ();
});


dotest.add ('Method agreements.list', function (test) {
  toonapi.agreements.list (function (err, data) {
    var item = data && data[0];

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isExactly ('fail', 'data[0].city', item && item.city, 'AMSTERDAM')
      .done ();
  });
});


dotest.add ('Method agreements.update', function (test) {
  test ()
    .warn ('Not implemented')
    .done ();
});


dotest.add ('Method display.status', function (test) {
  toonapi.display.status (function (err, data) {
    var thermos = data && data.thermostatInfo;

    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.success', data && data.success, true)
      .isObject ('fail', 'data.thermostatInfo', thermos)
      .isNumber ('warn', 'data.thermostatInfo.currentTemp', thermos && thermos.currentTemp)
      .done ();
  });
});


dotest.add ('Method devices.list', function (test) {
  toonapi.devices.list (function (err, data) {
    var item = data && data[0];

    cache.devices = data;

    test (err)
      .isArray ('fail', 'data', data)
      .isNotEmpty ('fail', 'data', data)
      .isObject ('fail', 'data[0]', item)
      .isNotEmpty ('fail', 'data[0]', item)
      .isString ('fail', 'data[0].uuid', item && item.uuid)
      .done ();
  });
});


dotest.add ('Method devices.update', function (test) {
  toonapi.devices.update (cache.devices, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.success', data && data.success, true)
      .done ();
  });
});


dotest.add ('Method devices.device.get', function (test) {
  var item = cache.devices && cache.devices[0];

  toonapi.devices.device.get (item && item.uuid, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.uuid', data && data.uuid, item && item.uuid)
      .done ();
  });
});


dotest.add ('Method temperature.update - without scale', function (test) {
  test ()
    .warn ('Not implemented')
    .done ();
});


dotest.add ('Method oauth.getTokenFromPassword', function (test) {
  toonapi.oauth.getTokenFromPassword (function (err, data) {
    config.accessToken = data && data.access_token;
    config.refreshToken = data && data.refresh_token;
    toonapi = app (config);

    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.token_type', data && data.token_type, 'bearer')
      .done ();
  });
});


dotest.add ('Method oauth.refreshToken', function (test) {
  toonapi.oauth.refreshToken (config.refreshToken, function (err, data) {
    config.accessToken = data && data.access_token;
    config.refreshToken = data && data.refresh_token;

    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.token_type', data && data.token_type, 'bearer')
      .done ();
  });
});


dotest.add ('Method oauth.revokeToken', function (test) {
  toonapi.oauth.revokeToken (config.accessToken, function (err, data) {
    var revRefresh = data && data.revokedrefreshtoken;
    var revAccess = data && data.revokedaccesstoken;

    test (err)
      .isObject ('fail', 'data', data)
      .isExactly ('fail', 'data.revokedrefreshtoken', revRefresh, config.refreshToken)
      .isExactly ('fail', 'data.revokedaccesstoken', revAccess, config.accessToken)
      .done ();
  });
});


dotest.add ('Method oauth.getTokenFromCode', function (test) {
  test ()
    .warn ('Not implemented')
    .done ();
});


dotest.add ('Method oauth.getTokenImplicit', function (test) {
  test ()
    .warn ('Not implemented')
    .done ();
});


dotest.add ('Error: API error - xml <fault>', function (test) {
  config.endpoint = 'https://api.toonapi.com/api';
  toonapi = app (config);

  toonapi.agreements.list (function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'API error')
      .isNumber ('fail', 'err.statusCode', err && err.statusCode)
      .isObject ('fail', 'err.error', err && err.error)
      .isNotEmpty ('fail', 'err.error', err && err.error)
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Error: timeout', function (test) {
  config.endpoint = 'https://api.toonapi.com';
  config.timeout = 1;
  toonapi = app (config);

  toonapi.agreements.list (function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.code', err && err.code, 'TIMEOUT')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


dotest.add ('Error: endpoint missing', function (test) {
  config.endpoint = '-';
  config.timeout = 5000;
  toonapi = app (config);

  toonapi.agreements.list (function (err, data) {
    test ()
      .isError ('fail', 'err', err)
      .isExactly ('fail', 'err.message', err && err.message, 'endpoint missing')
      .isUndefined ('fail', 'data', data)
      .done ();
  });
});


dotest.run ();


lib/oauth.js
var app;


/**
 * Get an access token from username and password
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function getTokenFromPassword (callback) {
  var options = {
    method: 'POST',
    noauth: true,
    path: '/token',
    params: {
      grant_type: 'password',
      client_id: app.config.clientId,
      client_secret: app.config.clientSecret,
      username: app.config.username,
      password: app.config.password
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Refresh access token
 *
 * @callback callback
 * @param token - Access token to refresh
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function refreshToken (token, callback) {
  var options = {
    method: 'POST',
    noauth: true,
    path: '/token',
    params: {
      client_id: app.config.clientId,
      client_secret: app.config.clientSecret,
      grant_type: 'refresh_token',
      refresh_token: token
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Revoke access token
 *
 * @callback callback
 * @param token - Access token to revoke
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function revokeToken (token, callback) {
  var options = {
    method: 'POST',
    noauth: true,
    path: '/revoke',
    params: {
      token: token,
      client_id: app.config.clientId,
      client_secret: app.config.clientSecret
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Set up lib
 *
 * @param parent {object} - Main module utils
 * @returns {object} - Interface methods
 */

function setup (parent) {
  app = parent;

  return {
    getTokenFromPassword,
    refreshToken,
    revokeToken
  };
}

module.exports = setup;


lib/devices.js
var app;


/**
 * Get device details
 *
 * @callback callback
 * @param deviceId {string} - Device UUID
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function deviceGet (deviceId, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/devices/' + deviceId
  };

  app.httpRequest (options, callback);
}


/**
 * Update device details
 *
 * @callback callback
 * @param deviceId {string} - Device UUID
 * @param obj {object} - Use and alter full object from `deviceGet()`
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function deviceUpdate (deviceId, obj, callback) {
  var options = {
    method: 'PUT',
    path: '/toon/api/v1/devices/' + deviceId,
    json: obj
  };

  app.httpRequest (options, callback);
}


/**
 * Get device data
 *
 * @callback callback
 * @param deviceId {string} - Device UUID
 * @param [params] {object} - Filter data
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function deviceData (deviceId, params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get device flows data
 *
 * @callback callback
 * @param deviceId {string} - Device UUID
 * @param [params] {object} - Filter data
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function deviceFlows (deviceId, params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get all devices array and their details
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function listAll (callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/devices'
  };

  app.httpRequest (options, callback);
}


/**
 * Update all devices at once
 *
 * @callback callback
 * @param arr {array} - Use and alter full array from `list()`
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function updateAll (arr, callback) {
  var options = {
    method: 'PUT',
    path: '/toon/api/v1/devices',
    json: arr
  };

  app.httpRequest (options, callback);
}


/**
 * Set up lib
 *
 * @param parent {object} - Main module utils
 * @returns {object} - Interface methods
 */

function setup (parent) {
  app = parent;

  return {
    device: {
      get: deviceGet,
      update: deviceUpdate,
      data: deviceData,
      flows: deviceFlows
    },
    list: listAll,
    update: updateAll
  };
}

module.exports = setup;


lib/pushEvent.js
var app;


/**
 * Subscribe a push event request
 *
 * @callback callback
 * @param callbackUri {string} - Callback URL
 * @param applicationId {string} - Application ID
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function subscribe (applicationId, callbackUrl, callback) {
  var options = {
    method: 'POST',
    path: '/toon/api/v1/pushEvent',
    json: {
      applicationId: applicationId,
      callbackUrl: callbackUrl
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Unsubscribe a push event request
 *
 * @callback callback
 * @param applicationId {string} - Application ID
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function unsubscribe (applicationId, callback) {
  var options = {
    method: 'DELETE',
    path: '/toon/api/v1/pushEvent',
    params: {
      applicationId: applicationId
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Set up lib
 *
 * @param parent {object} - Main module utils
 * @returns {object} - Interface methods
 */

function setup (parent) {
  app = parent;

  return {
    subscribe,
    unsubscribe
  };
}

module.exports = setup;


lib/consumption.js
var app;


/**
 * Get electricity data
 *
 * @callback callback
 * @param [params] {object} - Filter options
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param [params.interval] {string} - Hours, Days, Weeks, Months, Years
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function electricityData (params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/consumption/electricity/data'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get electricity flows data
 *
 * @callback callback
 * @param [params] {object} - Filter options
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function electricityFlows (params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/consumption/electricity/flows'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get districtheat data
 *
 * @callback callback
 * @param [params] {object} - Filter options
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param [params.interval] {string} - Hours, Days, Weeks, Months, Years
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function districtheatData (params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/consumption/districtheat/data'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get gas data
 *
 * @callback callback
 * @param [params] {object} - Filter options
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param [params.interval] {string} - Hours, Days, Weeks, Months, Years
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function gasData (params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/consumption/gas/data'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Get gas flows data
 *
 * @callback callback
 * @param [params] {object} - Filter options
 * @param [params.fromTime] {number} - Timestamp in ms
 * @param [params.toTime] {number} - Timestamp in ms
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function gasFlows (params, callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/consumption/gas/flows'
  };

  if (typeof params === 'function') {
    callback = params;
    params = null;
  }

  options.params = params;
  app.httpRequest (options, callback);
}


/**
 * Set up lib
 *
 * @param parent {object} - Main module utils
 * @returns {object} - Interface methods
 */

function setup (parent) {
  app = parent;

  return {
    electricity: {
      data: electricityData,
      flows: electricityFlows
    },
    districtheat: {
      data: districtheatData
    },
    gas: {
      data: gasData,
      flows: gasFlows
    }
  };
}

module.exports = setup;


lib/temperature.js
var app;


/**
 * Get temperature states array from Toon
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function statesList (callback) {
  var options = {
    method: 'GET',
    path: '/toon/api/v1/states'
  };

  app.httpRequest (options, callback);
}


/**
 * Change temperature state on Toon
 *
 * @callback callback
 * @param temperatureState {number} - manual (0), program (1)
 * @param state {number, string} - comfort (0), home (1), sleep (2), away (3)
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function statesUpdate (temperatureState, state, callback) {
  var tempStates = {
    manual: 0,
    program: 1
  };

  var states = {
    comfort: 0,
    home: 1,
    sleep: 2,
    away: 3
  };

  var options = {
    method: 'PUT',
    path: '/toon/api/v1/states',
    json: {
      temperatureState: tempStates [temperatureState] || temperatureState,
      state: states [state] || state
    }
  };

  app.httpRequest (options, callback);
}


/**
 * Set up lib
 *
 * @param parent {object} - Main module utils
 * @returns {object} - Interface methods
 */

function setup (parent) {
  app = parent;

  return {
    states: {
      list: statesList,
      update: statesUpdate
    }
  };
}

module.exports = setup;


index.js
var httpreq = require ('httpreq');

var app = {
  root: __dirname,
  config: {
    accessToken: null,
    clientId: null,
    clientSecret: null,
    endpoint: 'https://api.toonapi.com',
    redirectUrl: null,
    timeout: 5000
  }
};

// Load methods
app.oauth = require (app.root + '/lib/oauth.js') (app);
app.agreements = require (app.root + '/lib/agreements.js') (app);
app.consumption = require (app.root + '/lib/consumption.js') (app);
app.devices = require (app.root + '/lib/devices.js') (app);
app.display = require (app.root + '/lib/display.js') (app);
app.pushEvent = require (app.root + '/lib/pushEvent.js') (app);
app.temperature = require (app.root + '/lib/temperature.js') (app);


/**
 * Callback an error
 *
 * @callback callback
 * @param msg {string} - Error.message
 * @param err {Error, mixed} - Error.error
 * @param code {number} - Error.statusCode
 * @param callback {function} - `function (err) {}`
 * @returns {void}
 */

app.doError = function doError (msg, err, res, callback) {
  var error = new Error (msg);
  var body = res && res.body || '';

  function doCallback () {
    callback && callback (error);
  }

  // HTTP response error
  if (msg === 'invalid response' || msg === 'API error') {
    error.statusCode = res.statusCode;
    error.headers = res.headers;

    // API error - json
    if (!(err instanceof Error)) {
      error.error = err;
      doCallback ();
      return;
    }

    // API error - xml <fault>
    if (body.match (/^<(amt:|ams:)?fault/)) {
      error.message = 'API error';

      body.replace (/<(amt:|ams:)?fault [^>]+>(<\1code>([^<]+)<\/\1code>)?(<\1message>([^<]+)<\/\1message>)?(<\1description>([^<]+)<\/\1description>)?<\/\1fault>$/, function (str0, str1, str2, code, str4, message, str6, description) {
        error.error = {
          code,
          message,
          description
        };
      });

      doCallback ();
      return;
    }

    error.body = body;
    doCallback ();
    return;
  }

  // Normal error
  error.error = err;
  doCallback ();
};


/**
 * Process HTTP response for callback
 *
 * @callback callback
 * @param err {Error, null} - httpreq error
 * @param res {object} - httpreq response details
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function httpResponse (err, res, callback) {
  var data = res && res.body || '';

  if (err) {
    return callback (err);
  }

  // Parse response
  if (!data && res.statusCode === 200) {
    data = '{ "success": true }';
  }

  try {
    data = JSON.parse (data);
  } catch (e) {
    e.body = data;
    return app.doError ('invalid response', e, res, callback);
  }

  // Catch JSON API error
  if (res.statusCode >= 300) {
    return app.doError ('API error', data, res, callback);
  }

  return callback (null, data);
}


/**
  * Send HTTP request to the remote API
  *
  * @callback callback
  * @param props {object} - Request details
  * @param props.path {string} - Request path after `endpoint`
  * @param [props.method = GET] {string} - GET, POST, etc
  * @param [props.params] {object} - GET or POST fields to send
  * @param callback {function} - `function (err, data) {}`
  * @returns {void}
  */

app.httpRequest = function httpRequest (props, callback) {
  var options = {
    method: props.method || 'GET',
    url: app.config.endpoint + props.path,
    parameters: props.params,
    json: props.json,
    timeout: props.timeout || app.config.timeout,
    headers: props.headers || {}
  };

  if (!app.config.endpoint || !app.config.endpoint.match (/^https?:\/\/[\w\d\.]+/)) {
    return app.doError ('endpoint missing', null, null, callback);
  }

  // we need but don't have an accessToken
  if (!props.noauth && !app.config.accessToken) {
    app.oauth.getTokenFromPassword (function (err, token) {
      if (err) {
        return callback (err);
      }

      app.config.accessToken = token.access_token;
      app.config.refreshToken = token.refresh_token;

      app.httpRequest (props, callback);
    });

    return;
  }

  // we need and have an accessToken
  if (!props.noauth && app.config.accessToken) {
    options.headers.Authorization = 'Bearer ' + app.config.accessToken;
  }

  function doResponse (err, res) {
    if (props.path === '/revoke') {
      if (res.statusCode === 200) {
        callback (null, {
          revokedaccesstoken: res.headers.revokedaccesstoken,
          revokedrefreshtoken: res.headers.revokedrefreshtoken
        });

        return;
      }
    }

    httpResponse (err, res, callback);
  }

  options.headers.Accept = 'application/json';
  options.headers ['User-Agent'] = 'node/toonapi (https://github.com/fvdm/nodejs-toonapi)';
  httpreq.doRequest (options, doResponse);

  return null;
};


/**
 * Module config and interface
 *
 * @param set {object} - Configuration settings
 * @param set.clientId {string} - API client_id
 * @param set.clientSecret {string} - API client_secret
 * @param set.redirectUrl {string} - API redirect URL
 * @returns {object} - Interface methods
 */

function setup (set) {
  if (set && set instanceof Object) {
    app.config.accessToken = set.accessToken;
    app.config.clientId = set.clientId;
    app.config.clientSecret = set.clientSecret;
    app.config.endpoint = set.endpoint || app.config.endpoint;
    app.config.password = set.password;
    app.config.redirectUri = set.redirectUri;
    app.config.refreshToken = set.refreshToken;
    app.config.timeout = set.timeout || app.config.timeout;
    app.config.username = set.username;
  }

  return app;
}

module.exports = setup;
